File: src/main/java/com/lessmarkup/interfaces/data/QueryBuilder.java
package com.lessmarkup.interfaces.data;

import java.util.Collection;
import java.util.List;

public interface QueryBuilder {
    <T extends DataObject> QueryBuilder from(Class<T> type, String name);
    default <T extends DataObject> QueryBuilder from(Class<T> type) {
        return from(type, null);
    }
    <T extends DataObject> QueryBuilder join(Class<T> type, String name, String on);
    <T extends DataObject> QueryBuilder leftJoin(Class<T> type, String name, String on);
    <T extends DataObject> QueryBuilder rightJoin(Class<T> type, String name, String on);
    QueryBuilder where(String filter, Object ... args);
    QueryBuilder whereId(Long id);
    QueryBuilder whereIds(Collection<Long> ids);
    QueryBuilder orderBy(String column);
    QueryBuilder orderByDescending(String column);
    QueryBuilder groupBy(String column);
    QueryBuilder limit(int from, int count);
    <T extends DataObject> T find(Class<T> type, long id);
    <T extends DataObject> T findOrDefault(Class<T> type, long id);
    <T extends DataObject> List<T> execute(Class<T> type, String sql, Object ... args);
    boolean executeNonQuery(String sql, Object ... args);
    <T> T executeScalar(Class<T> type, String sql, Object ... args);
    default <T> List<T> toList(Class<T> type) {
        return toList(type, null);
    }
    <T> List<T> toList(Class<T> type, String selectText);
    List<Long> toIdList();
    int count();
    <T> T first(Class<T> type, String selectText);
    default <T> T first(Class<T> type) {
        return first(type, null);
    }
    <T> T firstOrDefault(Class<T> type, String selectText);
    default <T> T firstOrDefault(Class<T> type) {
        return firstOrDefault(type, null);
    }
    QueryBuilder createNew();
    <T extends DataObject> boolean deleteFrom(Class<T> type, String filter, Object ... args);
}

File: src/main/scala/com/lessmarkup/interfaces/data/QueryBuilder.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.interfaces.data

trait QueryBuilder {

  def from[T <: DataObject](dataType: Class[T], name: Option[String]): QueryBuilder

  def fromJava[T <: DataObject](dataType: Class[T], name: String): QueryBuilder = from(dataType, Option(name))

  def from[T <: DataObject](dataType: Class[T]) = fromJava(dataType, null)

  def join[T <: DataObject](dataType: Class[T], name: String, on: String): QueryBuilder

  def leftJoin[T <: DataObject](dataType: Class[T], name: String, on: String): QueryBuilder

  def rightJoin[T <: DataObject](dataType: Class[T], name: String, on: String): QueryBuilder

  def where(filter: String, args: Any*): QueryBuilder

  def whereId(id: Long): QueryBuilder

  def whereIds(ids: Seq[Long]): QueryBuilder

  def orderBy(column: String): QueryBuilder

  def orderByDescending(column: String): QueryBuilder

  def groupBy(column: String): QueryBuilder

  def limit(from: Int, count: Int): QueryBuilder

  def find[T <: DataObject](dataType: Class[T], id: Long): Option[T]

  def execute[T <: DataObject](dataType: Class[T], sql: String, args: Any*): Seq[T]

  def executeNonQuery(sql: String, args: Any*): Boolean

  def executeScalar[T](dataType: Class[T], sql: String, args: Any*): Option[T]

  def toList[T <: AnyRef](dataType: Class[T], selectText: Option[String]): Seq[T]

  def toList[T <: AnyRef](dataType: Class[T]): Seq[T] = toList(dataType, None)

  def toIdList: Seq[Long]

  def count: Int

  def first[T <: AnyRef](dataType: Class[T], selectText: Option[String]): Option[T]

  def createNew: QueryBuilder

  def deleteFrom[T <: DataObject](dataType: Class[T], filter: String, args: Any*): Boolean
}
